File: PokedexSprintChallengeObjC/PokedexSprintChallengeObjC/PokemonTableViewController.m
//
//  PokemonTableViewController.m
//  PokedexSprintChallengeObjC
//
//  Created by Dillon P on 5/3/20.
//  Copyright © 2020 Dillon's Lambda Team. All rights reserved.
//

#import "PokemonTableViewController.h"
#import "PokedexSprintChallengeObjC-Swift.h"
#import "Pokemon.h"

@interface PokemonTableViewController () <UISearchResultsUpdating, UISearchBarDelegate>

@property (nonatomic) UISearchController *searchController;
@property (nonatomic) NSArray<Pokemon *> *filteredPokemon;
@property (nonatomic) NSInteger selectedScopeButtonIndex;

@end

@implementation PokemonTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.searchController = [[UISearchController alloc] initWithSearchResultsController:nil];
    self.searchController.searchResultsUpdater = self;
    self.searchController.searchBar.delegate = self;
    self.searchController.obscuresBackgroundDuringPresentation = NO;
    self.searchController.searchBar.placeholder = @"Search Pokemon";
    self.searchController.searchBar.showsScopeBar = YES;
    self.searchController.searchBar.scopeButtonTitles = @[@"Pokemon Name", @"Pokemon Type"];
    self.navigationItem.searchController = self.searchController;
    self.definesPresentationContext = YES;
    
    self.controller = PokemonController.sharedController;
    [self.controller getAllPokemonByTypeWithCompletion:^(NSMutableArray * _Nullable allPokemon, NSError * _Nullable error) {
        if (error) {
            NSLog(@"%@", error);
        }

        if (allPokemon) {
            NSSortDescriptor *sortDescriptor = [[NSSortDescriptor alloc] initWithKey:@"name" ascending:YES];
            NSArray *sortedArray = [allPokemon sortedArrayUsingDescriptors:@[sortDescriptor]];
            self.allPokemon = sortedArray;
            self.filteredPokemon = sortedArray;
            [self.tableView reloadData];
        }
    }];
}

#pragma mark - Table view data source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.filteredPokemon.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"pokemonCell" forIndexPath:indexPath];
    Pokemon *pokemon = self.filteredPokemon[indexPath.row];
    
    cell.textLabel.text = [pokemon.name capitalizedString];
    
    return cell;
}


#pragma mark - Navigation

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"ShowPokemonDetailSegue"]) {
        PokemonDetailViewController *detailVC = segue.destinationViewController;
        
        NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
        Pokemon *pokemon = self.filteredPokemon[indexPath.row];
        
        detailVC.pokemon = pokemon;
        detailVC.controller = self.controller;
    }
}

#pragma mark - Search results updating delegate methods

- (void)updateSearchResultsForSearchController:(UISearchController *)searchController
{
    NSString *searchText = searchController.searchBar.text;
    if (searchText) {
        if (searchText.length != 0) {
            [self filterPokemonForSearchText:searchText];
        }
        else {
            self.filteredPokemon = self.allPokemon;
        }
        [self.tableView reloadData];
    }
}

- (void)searchBar:(UISearchBar *)searchBar selectedScopeButtonIndexDidChange:(NSInteger)selectedScope
{
    self.selectedScopeButtonIndex = selectedScope;
    [self updateSearchResultsForSearchController:self.searchController];
}

#pragma mark - Search bar helper methods


- (void)filterPokemonForSearchText:(NSString *)searchText
{
    NSString *filter = @"%K CONTAINS[cd] %@";
    NSPredicate *predicate = [NSPredicate new];
    if (self.selectedScopeButtonIndex == 0) {
        predicate = [NSPredicate predicateWithFormat:filter, @"name", searchText];
    } else if (self.selectedScopeButtonIndex == 1) {
        predicate = [NSPredicate predicateWithFormat:filter, @"type", searchText];
    }
    
    self.filteredPokemon = [self.allPokemon filteredArrayUsingPredicate:predicate];
}

@end

File: PokedexHybrid/PokedexHybrid/PokemonTableViewController.swift
//
//  PokemonTableViewController.swift
//  PokedexHybrid
//
//  Created by John Kouris on 2/22/20.
//  Copyright © 2020 John Kouris. All rights reserved.
//

import UIKit

class PokemonTableViewController: UITableViewController {
    
    private let reuseIdentifier = "PokemonCell"
    private let detailSegue = "ShowPokemonDetailSegue"
    
    var allPokemon = [Pokemon]()

    override func viewDidLoad() {
        super.viewDidLoad()
        fetchPokemon()
    }

    // MARK: - Table view data source

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return allPokemon.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier, for: indexPath)
        
        let pokemon = allPokemon[indexPath.row]
        cell.textLabel?.text = pokemon.pokemonName.capitalized
        
        return cell
    }

    // MARK: - Navigation

    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == detailSegue {
            guard let destinationVC = segue.destination as? PokemonDetailViewController,
                let indexPath = tableView.indexPathForSelectedRow else { return }
            destinationVC.pokemon = allPokemon[indexPath.row]
        }
    }
    
    // MARK: - Private methods
    
    private func fetchPokemon() {
        PokemonController.shared.fetchAllPokemon { (pokemon, error) in
            if let error = error {
                print("Error fetching pokemon: \(error)")
                return
            }
            
            guard let pokemon = pokemon else {
                print("No pokemon")
                return
            }
            
            self.allPokemon.append(contentsOf: pokemon)
            DispatchQueue.main.async {
                self.tableView.reloadData()
            }
        }
    }

}
